Add Spring Boot integration

sunny-elastic opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR.
-->

- Generated the skeleton of Spring boot endpoints integration package.
- Added 5 data streams (Jolokia, Prometheus, Info, Httptrace, Auditevents)
- Added data collection logic for all the data stream.
- Added the ingest pipeline for all the data stream.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^7.13.0`).

<!-- ## Author's Checklist

Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved

- [ ]
-->
## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/spring_boot_endpoints directory.
- Run the following command to run tests. 
> elastic-package test 


## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

endorama commented on packages/spring_boot_endpoints/docs/README.md:
```diff
@@ -0,0 +1,5 @@
+# Spring Boot Endpoints
+
+This is a new integration created using the [elastic-package](https://github.com/elastic/elastic-package) tool.
```
Please provide a more complete `README`. As an example you may refer to [Apache integration README](https://github.com/elastic/integrations/blob/main/packages/apache/_dev/build/docs/README.md)

sunny-elastic replied:
Yes sure PR is in progress. Expect Readme addition in later commits. Thanks!

endorama reviewed (changes requested):
I'm testing this locally, meanwhile I noticed the README is missing.

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2812/detail/PR-2812/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2812/detail/PR-2812/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2812/detail/PR-2812/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2812/detail/PR-2812/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_2812.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-03-10T08:53:11.556Z&rangeTo=2022-03-10T09:13:11.556Z&transactionName=BUILD Ingest-manager/integrations/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=175fe2f8507595dfe5bf1c2ffa625560&transactionId=66ae6c572251095c)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-03-10T09:03:11.556+0000


* Duration: 89 min 54 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 4251  |
| Skipped      | 7 |
| Total        | 4258   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2812/detail/PR-2812/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test integration: spring_boot_endpoints`
<ul>
<li>Took 1 min 36 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-2812/runs/1/steps/11088/log/?start=0">here</a></li>
<li>Description: <code>                                eval "$(../../build/elastic-package stack shellinit)"                                ../../build/elastic-package test -v --report-format xUnit --report-output file --test-coverage                                </code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mtojek commented on packages/spring_boot_endpoints/_dev/deploy/docker/docker-compose.yml:
```diff
@@ -0,0 +1,7 @@
+version: '2.3'
+services:
+  spring_boot_endpoints:
+    build: .
```
Do we need to submit a 20MB jar file to the repository? you should rather submit a java code, build it in Docker and run it later.

sunny-elastic replied:
Yes its done. We have updated accordingly. Thanks!

mtojek commented on packages/spring_boot_endpoints/manifest.yml:
```diff
@@ -0,0 +1,84 @@
+  - name: spring_boot_endpoints
+    title: Spring Boot Endpoints Metrics
+    description: Collect Spring Boot Endpoints metrics.
+    inputs:
```
I don't think this is the right approach. I would rather split these inputs based on the value they gave, not the channel (Jolokia, Prometheus, httpjson).

What kind of metrics will this integration be able to collect?

sunny-elastic replied:
This integration will be able to collect logs via httpjson and metrics from jolokia and prometheus module respectively.
https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-prometheus-collector.html
https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-jolokia-jmx.html"

mtojek commented on packages/spring_boot_endpoints/data_stream/prometheus/fields/fields.yml:
```diff
@@ -0,0 +1,114 @@
+              type: keyword
+            - name: uri
+              type: keyword
+        - name: metrics
```
This is wrong too.

You should try at least to split these metrics into smaller groups (http_server, jvm, process, gcc).

endorama replied:
Are there some guidelines to follow about this? I think it's hard to draw an objective line between "right" and "wrong" at the moment. I agree smaller groups would be preferred but it's difficult to understand what changes would be required to make this "right".

mtojek replied:
I think we should preserve consistency with the approach we follow in Beats. Look at this [fields.yml](https://github.com/elastic/beats/blob/main/metricbeat/module/apache/status/_meta/fields.yml) for Apache. Metrics are divided based on the area they belong (CPU, connections, scoreboard). It's the natural form, and I'm not sure if we have ever documented it.

sunny-elastic replied:
Ok sure. We will update it and split metrics accordingly.

mtojek reviewed (changes requested):
BTW we're working on the fix to prevent triggering too many jobs for a new package: https://github.com/elastic/integrations/pull/2814. Feel free to rebase once it's merged.

So summing up the first review round:
1. Please rename the package to Spring Boot, as there isn't the Spring Boot Endpoints product.
2. We should divide metrics into logical groups like (JVM, cluster, status, disk, etc.) than jolokia/jvm/etc.
3.

endorama commented on packages/spring_boot_endpoints/manifest.yml:
```diff
@@ -0,0 +1,84 @@
+format_version: 1.0.0
+name: spring_boot_endpoints
+title: Spring Boot Endpoints
+version: 0.0.1
```
Version `0.0.1` doesn't have a changelog entry, but version `0.1.0` does. Should this be like this instead?
```suggestion
version: 0.1.0
```

sunny-elastic replied:
yes its updated. Thanks!

sunny-elastic commented:
> BTW we're working on the fix to prevent triggering too many jobs for a new package: #2814. Feel free to rebase once it's merged.
> 
> So summing up the first review round:
> 
> 1. Please rename the package to Spring Boot, as there isn't the Spring Boot Endpoints product.
> 2. We should divide metrics into logical groups like (JVM, cluster, status, disk, etc.) than jolokia/jvm/etc.

1. We used the name "Sprint Boot Endpoints" since that is what the Github issue says: [https://github.com/elastic/integrations/issues/2569](https://www.google.com/url?q=https://github.com/elastic/integrations/issues/2569&sa=D&source=editors&ust=1646997159826935&usg=AOvVaw1a698iZGqNH8DE3aqAbC3F)
2. We thought of merging it in one to make it extensible(for future support of more metrics) and also referring to Hadoop integration PR. Is our understanding wrong?

sunny-elastic commented on packages/spring_boot_endpoints/manifest.yml:
```diff
@@ -0,0 +1,84 @@
+format_version: 1.0.0
+name: spring_boot_endpoints
+title: Spring Boot Endpoints
```
We used the name "Sprint Boot Endpoints" since that is what the Github issue we have followed with: [https://github.com/elastic/integrations/issues/2569](https://www.google.com/url?q=https://github.com/elastic/integrations/issues/2569&sa=D&source=editors&ust=1646997159826935&usg=AOvVaw1a698iZGqNH8DE3aqAbC3F)

mtojek commented:
> We thought of merging it in one to make it extensible(for future support of more metrics) and also referring to Hadoop integration PR. Is our understanding wrong?

Yes, I'm afraid it's wrong. You can always add more metrics to existing data streams. It's more natural to the end-users to enable metric areas they're interested in.

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

sunny-elastic commented:
> > We used the name "Sprint Boot Endpoints" since that is what the Github issue says: [https://github.com/elastic/integrations/issues/2569](https://www.google.com/url?q=https://github.com/elastic/integrations/issues/2569&sa=D&source=editors&ust=1646997159826935&usg=AOvVaw1a698iZGqNH8DE3aqAbC3F)
> 
> I'm afraid that's a mistake and we need to correct this.
> 
Ok. We follow the PRD as our primary source of document for any of our integrations to work on, thats why it was little confusion for us, but yeah will make this change and will change the name to **Spring Boot** in our next commit.
> > We thought of merging it in one to make it extensible(for future support of more metrics) and also referring to Hadoop integration PR. Is our understanding wrong?
> 
> Yes, I'm afraid it's wrong. You can always add more metrics to existing data streams. It's more natural to the end-users to enable metric areas they're interested in.
> 
> EDIT:
> 
> We accept the fact that the integration will not collect "every" metric of the observed product. We'd rather extend the data stream with new field definitions, with the right fields and descriptions than have an "open gate" to collect all metrics.

Yes agreed on the approach you want to suggest and we have even started making changes accordingly, just that there are few points to be discussed over same, and along with that will try to close it accordingly in the subsequent commits.

mtojek commented on packages/spring_boot/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## Requirements
+
+In order to ingest data from Spring Boot :
```
nit: Boot: (no space before the colon)

sunny-elastic replied:
Done. Its corrected

mtojek commented on packages/spring_boot/_dev/build/docs/README.md:
````diff
@@ -0,0 +1,81 @@
+</dependency>
+```
+- For access of jolokia add below dependency in `pom.xml` of Spring Boot Application.
+```sh
````
sh?

sunny-elastic replied:
Done. Its corrected

mtojek commented on packages/spring_boot/data_stream/prometheus/manifest.yml:
```diff
@@ -0,0 +1,73 @@
+type: metrics
+release: beta
+streams:
+  - input: prometheus/metrics
+    template_path: metrics.yml.hbs
+    title: Spring Boot Prometheus Metrics
+    description: Collect Spring Boot Prometheus Metrics.
```
Folks, I thought [we agreed](https://github.com/elastic/integrations/pull/2812#issuecomment-1068250904) on not using sink-like data streams, that aggregate metrics from many areas. Please re-iterate on this (and on Jolokia data stream) and propose more accurate metrics data streams instead.

mtojek replied:
@yug-elastic Could you please work together with Sunny and identify logical data streams? I remember you have done it right [here](https://github.com/elastic/integrations/pull/2811#issuecomment-1067843042).

yug-elastic replied:
Sure, @mtojek. We'll try to identify the logical divisions, get it reviewed by you and split the data streams accordingly.

sunny-elastic replied:
Based on the logical splitting of data streams 
Jolokia - 3 data streams [Resource, Cluster, Status]
Prometheus - 2 data streams [Java Memory, Server]
@mtojek let me know if it look fine or you want to suggest any other relevant names for data streams. Thanks!

mtojek replied:
First of all, you don't need to collect ALL available metrics. It's better to select a minimal set that gives user value than store everything and generates costs. Let me copy your slack post here:

```
For Jolokia we formed 3 data streams [Resource, Cluster, Status]
RESOURCE
	name=direct,type=BufferPool
	name=mapped,type=BufferPool
	type=Memory
	name=PS MarkSweep,type=GarbageCollector
	name=PS Old Gen,type=MemoryPool
	name=CodeCacheManager,type=MemoryManager
	name=Metaspace Manager,type=MemoryManager
	type=OperatingSystem
CLUSTER
	type=Threading
	type=Runtime
	type=Config
	type=ServerHandler
	type=Discovery
STATUS
	type=ClassLoading
	type=Compilation
	type=HotSpotDiagnostic
	type=DiagnosticCommand
For Prometheus we formed 2 data streams [Java Memeory, Server]
Java Memory
	All JVM related fields
Server
	All server related fields like http, tomcat ...etc
``` 

What do you think about different buckets, more similar to what we have in other integrations:
* gc (garbage collector, old gen, marksweep)
* memory (buffer pool, type=Memory?)
* os
* threading
* server (http, tomcat)

To be dropped:
* CodeCacheManager, Metaspace Manager (do they give any greater user value?)
* ClassLoading, Compilation, HotSpotDiagnostic, DiagnosticCommand (dropped as it seems to be noise?)

Could you please justify the `cluster` choice? What kind of metrics (with samples) does it collect?

sunny-elastic replied:
> First of all, you don't need to collect ALL available metrics. It's better to select a minimal set that gives user value than store everything and generates costs. Let me copy your slack post here:
> 
> ```
> For Jolokia we formed 3 data streams [Resource, Cluster, Status]
> RESOURCE
> 	name=direct,type=BufferPool
> 	name=mapped,type=BufferPool
> 	type=Memory
> 	name=PS MarkSweep,type=GarbageCollector
> 	name=PS Old Gen,type=MemoryPool
> 	name=CodeCacheManager,type=MemoryManager
> 	name=Metaspace Manager,type=MemoryManager
> 	type=OperatingSystem
> CLUSTER
> 	type=Threading
> 	type=Runtime
> 	type=Config
> 	type=ServerHandler
> 	type=Discovery
> STATUS
> 	type=ClassLoading
> 	type=Compilation
> 	type=HotSpotDiagnostic
> 	type=DiagnosticCommand
> For Prometheus we formed 2 data streams [Java Memeory, Server]
> Java Memory
> 	All JVM related fields
> Server
> 	All server related fields like http, tomcat ...etc
> ```
> 
> What do you think about different buckets, more similar to what we have in other integrations:
> 
> * gc (garbage collector, old gen, marksweep)
> * memory (buffer pool, type=Memory?)
> * os
> * threading
> * server (http, tomcat)
> 

Yeah this looks fine to me. Just that wil also add Prometheus data streams [jvm] 

> To be dropped:
> 
> * CodeCacheManager, Metaspace Manager (do they give any greater user value?)

yes we can drop this since not getting relevant data. 

> * ClassLoading, Compilation, HotSpotDiagnostic, DiagnosticCommand (dropped as it seems to be noise?)

Yes correct. will drop this as well 

> 



> Could you please justify the `cluster` choice? What kind of metrics (with samples) does it collect?

So if you see **Threading and Runtime** looks fine having relevant fields to capture but **config** has following fields in response 
HistorySize
HistoryMaxEntries
MaxDebugEntries
Debug

so this **config** also we can drop.

**ServerHandler & Discovery** is not relevant we can drop that since it has empty response.

So here is the final list of data streams - 
 
From Jolokia - 
 - gc (MarkSweep, type=GarbageCollector)
 - memory (type=BufferPool, type=Memory, Old Gen,type=MemoryPool)
 - os (type=OperatingSystem, type=Runtime)
 - threading (type=Threading)

From Prometheus -
 - jvm (all jvm related fields)
 - server (http, tomcat etc)

@mtojek please confirm on the same if it looks fine.

mtojek replied:
LGTM!

sunny-elastic commented on packages/spring_boot/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## Compatibility
+
+This module has been tested against `Spring Boot Version: 2.3.12`
```
Done

sunny-elastic commented on .github/CODEOWNERS:
```diff
@@ -109,6 +109,7 @@
 /packages/snyk @elastic/security-external-integrations
 /packages/sonicwall @elastic/security-external-integrations
 /packages/sophos @elastic/security-external-integrations
+/packages/spring_boot @elastic/integrations
```
yes its updated now. Thanks!

mtojek commented on .github/CODEOWNERS:
```diff
@@ -1,5 +1,5 @@
 # Default owners
-* @elastic/integrations-developer-experience
+* @elastic/ecosystem
```
Could you please rebase your PR against the main branch? or merge the main branch here.

ruflin commented:
@mtojek @sunny-elastic How close are we of getting this PR in. If we predict a few more iterations, I recommend to switch over to the per data stream approach https://github.com/elastic/integrations/pull/2811#issuecomment-1082701692

mtojek commented:
This PR contains ~120 files, so I wouldn't mind splitting it into smaller chunks to review :)

sunny-elastic commented:
> @mtojek @sunny-elastic How close are we of getting this PR in. If we predict a few more iterations, I recommend to switch over to the per data stream approach [#2811 (comment)](https://github.com/elastic/integrations/pull/2811#issuecomment-1082701692)

Hey @ruflin and @mtojek : I think we have already addressed all review comments and evrything is been reviewed so far in terms of this PR. I would suggest if all looks good then we can merge this PR. Though we wil split data stream wise PR's over next connectors from the beginning itself. 
So for this If you still facing challenges to review this PR and becomes difficult to take this forwards for merge then wil try to create multiple PR's, let me know

mtojek commented:
> I think we have already addressed all review comments and evrything is been reviewed so far in terms of this PR. 

To be honest, we haven't even started looking deeper into your changes. We stopped at the data stream segregation level.

> So for this If you still facing challenges to review this PR and becomes difficult to take this forwards for merge then wil try to create multiple PR's, let me know

It isn't about addressing the feedback or not. It's about the difficulty of going through the PR and possibly missing bugs. If it isn't a huge problem, I'd suggest opening a separate draft PR and disassembly this PR into smaller chunks (as @ruflin suggested for other PRs).

sunny-elastic commented:
> > I think we have already addressed all review comments and evrything is been reviewed so far in terms of this PR.
> 
> To be honest, we haven't even started looking deeper into your changes. We stopped at the data stream segregation level.
> 
> > So for this If you still facing challenges to review this PR and becomes difficult to take this forwards for merge then wil try to create multiple PR's, let me know
> 
> It isn't about addressing the feedback or not. It's about the difficulty of going through the PR and possibly missing bugs. If it isn't a huge problem, I'd suggest opening a separate draft PR and disassembly this PR into smaller chunks (as @ruflin suggested for other PRs).

Sure @mtojek , as we discussed over the channel, my concern is this integration contains total 9 data streams, so we have clubbed 3 data streams per PR and raised new 3 PR's. Hope that will help for better review process. I wil add the reviewers over those PR's once the pipelines are green. Thanks!

ruflin commented:
I thought we agree of 1 PR per data streams? Why did we group these now together?
